### PR TITLE
Add tables, emitter, flusher for group module

### PR DESCRIPTION
**Pull request by RogerKSI**
## Implementation details
- Add tables in Flusher for group module
- Add hasura tracking for tables of group module
- Add emitter and flusher logic

image

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: hasura/hasura-metadata/tables.yaml [hunk 1]
```diff
@@ -48,6 +48,34 @@
           table:
             name: deposits
             schema: public
+    - name: group_members
+      using:
+        foreign_key_constraint_on:
+          column: account_id
```
Comment (colmazia) [inline hasura/hasura-metadata/tables.yaml:54]: No need to have a constrain on group_id column?

File: hasura/hasura-metadata/tables.yaml [hunk 2]
```diff
@@ -48,6 +48,34 @@
...
+          table:
+            name: group_members
+            schema: public
+    - name: group_policies
+      using:
+        foreign_key_constraint_on:
+          column: admin
```
Comment (colmazia) [inline hasura/hasura-metadata/tables.yaml:61]: This too?

File: hasura/hasura-metadata/tables.yaml [hunk 3]
```diff
@@ -48,6 +48,34 @@
...
+          table:
+            name: group_policies
+            schema: public
+    - name: group_votes
+      using:
+        foreign_key_constraint_on:
+          column: voter_id
```
Comment (colmazia) [inline hasura/hasura-metadata/tables.yaml:68]: group_proposal_id?

Comment (colmazia) [reply hasura/hasura-metadata/tables.yaml:68]: I think you are doing right, haha, you can treat this like a note where I don't understand

**Review by colmazia — COMMENTED** [review]
Have 1 fix and several questions

File: hasura/hasura-metadata/tables.yaml (hunk 1 above)
Comment (RogerKSI) [reply hasura/hasura-metadata/tables.yaml:54]: this is a relation under the account table. There is constrain on group_id in the group table.

File: hasura/hasura-metadata/tables.yaml (hunk 2 above)
Comment (RogerKSI) [reply hasura/hasura-metadata/tables.yaml:61]: ditto

File: flusher/flusher/db.py [hunk 4]
```diff
@@ -484,3 +484,74 @@ def Column(*args, **kwargs):
...
+    Column("total_weight", sa.String),
+    Column("created_at", CustomDateTime),
+)
+
+group_members = sa.Table(
+    "group_members",
+    metadata,
+    Column("group_id", sa.Integer, sa.ForeignKey("groups.id"), primary_key=True),
+    Column("account_id", sa.Integer, sa.ForeignKey("accounts.id"), primary_key=True),
+    Column("weight", sa.String),
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply flusher/flusher/db.py:509]: ok

**Review by colmazia — APPROVED** [review]

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/emitter/handler.go [hunk 5]
```diff
@@ -118,12 +124,53 @@ func (h *Hook) handleMsg(ctx sdk.Context, txHash []byte, msg sdk.Msg, log sdk.AB
...
+	prefix := []byte{groupkeeper.ProposalsByVotingPeriodEndPrefix}
+
+	iterator := ctx.KVStore(h.groupStoreKey).
+		Iterator(prefix, sdk.PrefixEndBytes(append(append(prefix, lenTimeByte), timeBytes...)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		proposalID, _ := splitKeyWithTime(iterator.Key())
+		h.doUpdateGroupProposal(ctx, proposalID)
+	}
+
```
[the remark this answers is not in the document]
Comment (colmazia) [reply hooks/emitter/handler.go:173]: ok

**Pull request closed without merging**